### PR TITLE
build,ci,release: add Linux x86_64 FIPS build

### PR DESCRIPTION
**Pull request by rail**
* Added a script to build   [golang-fips](https://github.com/golang-fips/go). The resulted Go version has a `fips` suffix, so we can easily select it in Bazel.
* Added a new Bazel toolchain and build config used by the FIPS build only.
* The FIPS docker image has FIPS-compliant `openssl` installed.
* Added a script to run FIPS build in CI.
* Tweaked one of the regexes to be less greedy.
* Added the platform to the release process.

Epic: DEVINF-478
Fixes: DEVINF-634
Fixes: DEVINF-633

Release note (build change): As a part of this change we start
publishing a FIPS compliant tarball and docker image for the Linux
x86_64 platform. The build uses OpenSSL libraries for crypto operations
by `dlopen`ing the corresponding dynamic libraries.

### CODE REVIEW

**Comment by rail** [comment]
This is just a draft that I'd like to get some feedback.

So far:
* I can build the toolchain. To simplify the process I used the same docker image RedHat uses to test there patches on.
* The new platform is not enabled in ci yet. Will need to poke around publish-artifacts and friends
* the rules_go will need some changes, maybe something [like this](https://gist.github.com/rail/593cd6a18d110e066f80e00e149fb950) (just hand waving for now).

File: pkg/release/build.go
Comment (rickystewart) [inline pkg/release/build.go:91]: Nit: can do `case PlatformLinux, PlatformLinuxFIPS:` instead of duplicating the arm.

File: build/teamcity/internal/release/build-and-publish-patched-go-fips.sh
Comment (rickystewart) [inline build/teamcity/internal/release/build-and-publish-patched-go-fips.sh:5]: Nit: Would be good to have this be part of the same job that builds the rest of the `go` toolchains so we don't have to trigger multiple jobs to upgrade `go`.

File: pkg/release/build.go
Comment (rickystewart) [inline pkg/release/build.go:247]: Nit: Do we want FIPS + ARM?

Comment (rail) [reply pkg/release/build.go:247]: Yes. I just started with one platform for now, because I need to fix the header search issue when we cross build arm64 on amd64.

File: build/teamcity/internal/release/build-and-publish-patched-go-fips.sh
Comment (rail) [reply build/teamcity/internal/release/build-and-publish-patched-go-fips.sh:5]: Agree. I'll merge the scripts when I'm happy with the result.

File: pkg/release/build.go
Comment (rail) [reply pkg/release/build.go]: We decided to go with intel only for now.

**Comment by rail** [comment]
This version works as expected. The only annoying thing is that we need to edit `.bazelrc` to specify the version when we update the toolchain. I wish there was a way to "tag" toolchains and select them based on the tag.

**Comment by rail** [comment]
If the PR looks good, I will need to:
* rebuild the Go toolchain, maybe using 1.19.4 for FIPS instead of current 1.19.5 (so we use the same version)
* Update the `download_go_sdk` entries for either both or just the FIPS entry.

**Comment by rickystewart** [comment]
> I wish there was a way to "tag" toolchains and select them based on the tag.

This is *exactly* how Bazel's [toolchain selection](https://bazel.build/extending/toolchains) works.

**Review by rickystewart — COMMENTED** [review]
Before we check this in can you add a check in CI to validate that the binaries produced do NOT have any FIPS stuff in them? My understanding is `go tool nm $BINARY | grep FIPS` should return no output if the binary is not FIPS-capable. The current behavior seems correct but it makes me a little anxious that `rules_go` is "accidentally" going to pick the FIPS toolchain when it shouldn't at some point. A check in `make-and-publish-build` to make sure the FIPS binary *does* actually have FIPS stuff would be good too.

Up to you whether we keep doing it this way or if you want to set up the toolchain the traditional way, but if it works, then it works. 🤷

**Comment by rail** [comment]
> Before we check this in can you add a check in CI to validate that the binaries produced do NOT have any FIPS stuff in them? My understanding is `go tool nm $BINARY | grep FIPS` should return no output if the binary is not FIPS-capable. The current behavior seems correct but it makes me a little anxious that `rules_go` is "accidentally" going to pick the FIPS toolchain when it shouldn't at some point. A check in `make-and-publish-build` to make sure the FIPS binary _does_ actually have FIPS stuff would be good too.

Let me implement these before merging this change.

**Comment by rail** [comment]
FTR. We chatted with Ricky and decided to take the current approach so we don't need to patch rules_go heavily. After I test the release process I will request a review for the resulted patch.

**Comment by rail** [comment]
This is ready for  review. I cannot fully test it for the 23.1 release, because there is one blocking test failure, that blocks Make and Publish. But I tested it before the blocker, so it should just work.
Additionally I tested the 22.2 backport and it worked as expected.

**Review by rickystewart — CHANGES_REQUESTED** [review]
Looks pretty good!

File: build/teamcity/cockroach/ci/builds/build_linux_x86_64_fips.sh
Comment (rail) [reply build/teamcity/cockroach/ci/builds/build_linux_x86_64_fips.sh:11]: Good point! Will do

**Comment by rail** [comment]
Finally, passed all flakes and got a [dry-run release](https://teamcity.cockroachdb.com/buildConfiguration/Internal_Release_PublishCockroachRelease/8739254?buildTab=artifacts). 
image

File: build/teamcity/cockroach/ci/builds/build_impl.sh
Comment (rickystewart) [inline build/teamcity/cockroach/ci/builds/build_impl.sh:45]: Nice.

**Review by rickystewart — APPROVED** [review]

**Comment by rail** [comment]
bors r=rickystewart

**Pull request merged**
